sdks/java/extensions/protobuf/src/main/java/org/apache/beam/sdk/extensions/protobuf/ProtoSchema.java

```diff
@@ -0,0 +1,326 @@
+          case "google.protobuf.UInt32Value":
+          case "google.protobuf.BytesValue":
+            return new ProtoRow.WrapperOverlay(this, protoField);
+          case "google.protobuf.Duration":
+          default:
+            if (isMap(protoField)) {
+              return new ProtoRow.MapOverlay(this, protoField);
+            } else {
+              return new ProtoRow.MessageOverlay(this, protoField);
+            }
+        }
+      case GROUP:
+      default:
+        throw new RuntimeException("Field type not matched.");
+    }
+    if (nullable) {
+      return new ProtoRow.NullableOverlay(protoField, fieldOverlay);
+    }
+    return fieldOverlay;
+  }
+
+  private ArrayList<ProtoRow.FieldOverlay> createFieldLayer(Descriptors.Descriptor descriptor) {
+    // Oneof fields are nullable, even as they are primitive or enums
+    List<Descriptors.FieldDescriptor> oneofMap = new ArrayList<>();
+    descriptor
+        .getOneofs()
+        .forEach(oneof -> oneof.getFields().forEach(field -> oneofMap.add(field)));
+
+    ArrayList<ProtoRow.FieldOverlay> fieldOverlays = new ArrayList<>();
+
+    Iterator<Descriptors.FieldDescriptor> protoFields = descriptor.getFields().iterator();
+    for (int i = 0; i < descriptor.getFields().size(); i++) {
+      Descriptors.FieldDescriptor protoField = protoFields.next();
+      if (protoField.isRepeated() && !isMap(protoField)) {
+        fieldOverlays.add(new ProtoRow.ArrayOverlay(this, protoField));
+      } else {
+        fieldOverlays.add(createFieldLayer(protoField, oneofMap.contains(protoField)));
+      }
+    }
+    return fieldOverlays;
+  }
+
+  private void init() {
+    this.getters = createFieldLayer(descriptor);
+
+    Schema.Builder builder = Schema.builder();
+    for (ProtoRow.FieldOverlay field : getters) {
+      builder.addField(field.getSchemaField());
+    }
+
+    schema = builder.build();
+
+    try {
+      if (DynamicMessage.class.equals(rawType)) {
+        this.fnNewBuilder = rawType.getMethod("newBuilder", Descriptors.Descriptor.class);
+      } else {
+        this.fnNewBuilder = rawType.getMethod("newBuilder");
+      }
+    } catch (NoSuchMethodException e) {
+    }
+  }
+
+  public Schema getSchema() {
+    return this.schema;
+  }
+
+  SchemaCoder<Message> getSchemaCoder() {
+    return SchemaCoder.of(schema, getToRowFunction(), getFromRowFunction());
+  }
+
+  private SerializableFunction<Message, Row> getToRowFunction() {
+    return new MessageToRowFunction(this);
+  }
+
+  private SerializableFunction<Row, Message> getFromRowFunction() {
+    return new RowToMessageFunction(this);
+  }
+
+  private void writeObject(ObjectOutputStream oos) throws IOException {
+    oos.defaultWriteObject();
+    if (DynamicMessage.class.equals(this.rawType)) {
+      if (this.descriptor == null) {
+        throw new RuntimeException("DynamicMessages require provider a Descriptor to the coder.");
+      }
+      ProtoDescriptorSerializer.writeObject(oos, this.descriptor);
```

Comment:
As mentioned above - just serialize this.descriptor.toProto() instead

Reply:
see ProtoDomain comment.